Added TFLite parsing

ShobhitBehl opened the pull request.

ShobhitBehl committed:
Add TFLite

Added code to parse a .tflite file into a Graph object,
extracting operator attributes onto nodes and extracting tensor
attrubtes onto edges. Added build rules for the same.

rachitagrawal commented on TFLite/src/tflite/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from os.path import dirname, basename, isfile, join
```
Some readability feedback: Never import the individual functions. Always import a package or a module. In this case, you can just import path.

Also, it's advisable to import it like from <package> import <module>, for better readability and consistency.

rachitagrawal commented on TFLite/src/tflite/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from os.path import dirname, basename, isfile, join
+import glob
+modules = glob.glob(join(dirname(__file__), "*.py"))
+__all__ = [ basename(f)[:-3] for f in modules if isfile(f) and not f.endswith('__init__.py')]
```
nit: Good idea to have comments in your own code.

rachitagrawal reviewed (changes requested):
I haven't gone through the entire PR. Just observed few things so mentioning those here:

1. Good idea to have one line description about the commit and then some details about it.
2. Remove the auto-generated files.
3. Another small description about the package in each BUILD file at the top of the file.

ShobhitBehl committed:
Bug fix and added init.py

Removed auto generated files, added a module(init.py) to generate
these files from schema.fbs using flatc compiler.
Added build rules for the same.

ShobhitBehl committed:
Update README.md

ShobhitBehl commented:
Removed auto-generated files and added code to create them using a bazel build.

ShobhitBehl committed:
Bug fix and update .gitignore

Fixed error in TFLite/src/init.py, updated .gitignore, added more
comments and fixed import format across files.

prakashprabhu commented on TFLite/src/Graph.py:
```diff
@@ -0,0 +1,105 @@
+
```
Are the Edge, Node, Graph classes specific to tflite or apply to other input types like graphdef also? In that case, it might be better to put it in a parent directory or subdirectory like 'common' or something similiar

ShobhitBehl replied:
Added a common sub-directory in src for the same

prakashprabhu commented on TFLite/src/Node.py:
```diff
@@ -0,0 +1,34 @@
+# Node module to store operator attributes
+class Node:
+    def __init__(self, label, value = None):
```
nit: A short comment on the semantics of each field might be helpful if it is not too time consuming.

ShobhitBehl replied:
Added a link in comments for semantic meaning and operator support

prakashprabhu commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,133 @@
+import flatbuffers
```
This looks great, I think finalizing the schema, and commiting to Spanner, running this to generate features for the hosted models on tflite and populating and write a few useful queries would be great to have as a e2e milestone for mid-point.

ShobhitBehl committed:
Added common sub-directory and comments

Added a common sub-directory to store generic modules Graph, Node and
Edge.
Added links in comments in Node and OpToNode for operator support and
semantic meaning of attributes
Added function to calculate duplication of operators and tensors
accross all files in models.

ShobhitBehl committed:
Remove some attributes from Node

Remove unecessary attributes from Node, update .gitgnore.

rachitagrawal commented on TFLite/src/main.py:
```diff
@@ -1,11 +1,51 @@
-    graph = tflite_parser.parse_graph("models/01062020-193418-e2.tflite")
+    get_duplication(tflite_parser)
+
+    graph = tflite_parser.parse_graph("models/smartreply.tflite")
```
May be instead of hardcoding the tflite model name, pass a CLA? argparse?

ShobhitBehl replied:
Committed a fix, now using argparse.

rachitagrawal commented on TFLite/src/init.py:
```diff
@@ -0,0 +1,27 @@
+def init_tflite_files():
+
+    # Absolute path to schema.fbs and the path to store the generated code
+    schema_path = '/home/shobhitbehl/ml-model-characterization/TFLite/schema.fbs'
```
Same here as well. These paths should be a command line arguments, instead of hard-coding them.

ShobhitBehl replied:
These files are fixed paths, not intended as user arguments. Changes the absolute path to relative path as a fix

rachitagrawal reviewed (approved):
Few minor changes suggested.

rachitagrawal commented on TFLite/src/common/Storage.py:
```diff
@@ -0,0 +1,208 @@
+# and _load_tensors implement batching
+class Storage:
+
+    # Initialize spanner client with instance and database
```
It's preferable to use docstrings for details about the class and class methods. But we can keep it like this for now. All future changes you can handle them accordingly.

ShobhitBehl replied:
Added docstrings

rachitagrawal commented on TFLite/src/init.py:
```diff
@@ -27,4 +22,9 @@ def init_tflite_files():
\ No newline at end of file
+
+    # Path to schema and path where the directory should be generated
+    schema_path = "./schema.fbs"
```
Make this a global constants? SCHEMA_PATH and OUT_PATH?

ShobhitBehl replied:
Changed to global constants

rachitagrawal commented on TFLite/src/main.py:
```diff
@@ -1,9 +1,14 @@
+import argparse
+
+import timeit
 
```
Let's keep two section of imports. Standard python library imports(os, argparse, timeit) and third-party imports like TFLiteParser etc). All of them should be lexicographically sorted.

ShobhitBehl replied:
Fixed imports

rachitagrawal replied:
Looks like all your updates are gone after the PR refactoring.

rachitagrawal commented on Queries.sql:
```diff
@@ -0,0 +1,83 @@
+SELECT
```
A description about each query would be helpful.

ShobhitBehl replied:
Added one line comment about each query

rachitagrawal commented on TF/src/OpToNode.py:
```diff
@@ -0,0 +1,72 @@
+            node.filter_height = operation.inputs[1].get_shape().as_list()[0]
+            node.filter_width = operation.inputs[1].get_shape().as_list()[1]
+        
+        elif op == "MaxPool" or op == "AvgPool":
```
Do we want to handle "ArgMaxPool" as well here?

ShobhitBehl replied:
MaxPoolWithArgmax in tf.raw_ops has slightly different attributes than MaxPool and AvgPool, haven't seen it in models yet, will add in subsequent iterations?

rachitagrawal commented on TF/src/main.py:
```diff
@@ -0,0 +1,50 @@
+import TFParser
+import argparse
+from common import Storage
```
Same as above.

ShobhitBehl replied:
Fixed

rachitagrawal reviewed (commented):
Have added some feedbacks.

rachitagrawal commented:
Let's not add more commits to this PR. Usually, we should create a new PR for any new task. Otherwise, the PR will be shown as approved even when reviewers have not approved all the commits. From now, let's create a new PR for a new task. To do the same, you can push your changes to a new branch into your github fork and from there create a new PR. Feel free to let me know if you face any problem.

rachitagrawal commented on TF/src/main.py:
```diff
@@ -30,21 +33,23 @@ def run_inference(filename, model_name, category, is_saved_model):
+    parser.add_argument('--model_name')
+    parser.add_argument('--category') 
+    parser.add_argument('--is_saved_model')
+    parser.add_argument('--input_operation_names', nargs="+", default = [])
```
Are all these arguments "required" argument? Can we add some default values for them?

ShobhitBehl replied:
Added default "True" to is_saved_model and "None" to category, filename and model_name are required.

rachitagrawal reviewed (commented):
Comments added.

rachitagrawal commented on common/Storage.py:
```diff
@@ -168,7 +168,7 @@ def _load_tensors(self, graph):
             # Number of edges to be processed per batch
             num_attributes = len(vars(graph.edges[0])) + 4
-            num_edges_per_batch = 200000 // num_attributes
+            num_edges_per_batch = 20000 // num_attributes
```
convert this constant to a constat literal?

ShobhitBehl replied:
Added a constant class attribute : _MAX_MUTATIONS.

rachitagrawal commented on TF/src/TFParser.py:
```diff
@@ -17,18 +17,19 @@ class TFParser:
             graph_def = meta_graph.graph_def
 
-            tf.io.write_graph(graph_def, "/home/shobhitbehl", "Test1.pb")
+            # tf.io.write_graph(graph_def, "/home/shobhitbehl/GraphDef", model_name + ".pb")
```
out_path for destination directory?

ShobhitBehl replied:
Removed the line, unnecessary code.

rachitagrawal commented on load_data.sh:
```diff
@@ -31,4 +30,8 @@ bazel-bin/tf_main --filename efficientnet_b5_feature-vector_1_ImageFeatureVector
-bazel-bin/tf_main --filename vision_embedder_inaturalist_V2_1_ImageFeatureVector_SavedModel.pb --model_name vision_embedder_inaturalist_V2_1 --category ImageFeatureVector --is_saved_model True
\ No newline at end of file
+bazel-bin/tf_main --filename vision_embedder_inaturalist_V2_1_ImageFeatureVector_SavedModel.pb --model_name vision_embedder_inaturalist_V2_1 --category ImageFeatureVector --is_saved_model True
+bazel-bin/tf_main --filename tweening_conv3d_bair_1_VideoGeneration_SavedModel.pb --model_name tweening_conv3d_bair_1 --category VideoGeneration --is_saved_model True 
```
is_saved_model default value can be 'True' as most of them are saved models.

ShobhitBehl replied:
Fixed

prakashprabhu commented on README.md:
```diff
@@ -1 +1,35 @@
+
+##  Load models into Spanner DB from scratch
+
+- Download the [data](https://drive.google.com/file/d/1DxtiCvxDLJ950g-2gd__Avu1dgupZLv6/view?usp=sharing)
```
This drive URL is publicly available? Is it possible to add it to the github repo?

ShobhitBehl replied:
Leaving this as is for now.

prakashprabhu commented on Schema_DDL:
```diff
@@ -1,6 +1,7 @@
 CREATE TABLE Models (
   model_name STRING(1024) NOT NULL,
   category STRING(1024),
+  source STRING(1024),
```
Does it make sense make source an ENUM? Since likely there would not be many sources, and this would also indicate what sources are currently supported?

ShobhitBehl replied:
source can only be "TF" or "TFLite", added comments for the same. More readable for querying.

prakashprabhu commented on TF/src/TFParser.py:
```diff
@@ -1,78 +1,101 @@
-    # Class instances to convert ops to nodes and tensors to edges
-    _OP_TO_NODE = OpToNode.OpToNode()
-    _TENSOR_TO_EDGE = TensorToEdge.TensorToEdge()
+    _NOT_OUTPUT = [
```
Is this list based on a general pattern that you identified? In that case, it might help to add a comment on how it was done.

ShobhitBehl replied:
Tentative list based on model analysis. Added comment for the same.

prakashprabhu commented on TF/src/main.py:
```diff
@@ -1,50 +1,76 @@
+
     tf_parser = TFParser.TFParser()
-    graph = tf_parser.parse_graph("models/" + filename, model_name, category, is_saved_model)
-    storage = Storage.Storage('ml-models-characterization-db', 'models_db')
```
Perhaps take the Spanner instance and db name as flags as well?

prakashprabhu commented on TFLite/src/OpToNode.py:
```diff
@@ -1,14 +1,53 @@
+        "ZEROS_LIKE" : "ZerosLike", "ABS" : "Abs", "HARD_SWISH" : "HardSwish",
+        "FILL" : "Fill", "REVERSE_V2" : "ReverseV2", "RANK" : "Rank",
+        "SEGMENT_SUM" : "SegmentSum", "SCATTER_ND" : "ScatterNd"
+    } # Mapping from TFLite to TF operations, not all ops have a one to one mapping
```
Can you comment on how you handle ops having many-to-1 or 1-to-many in that case?

ShobhitBehl replied:
Only 1-1 case handled, added comment to reflect the same.

prakashprabhu commented on TFLite/src/main.py:
```diff
@@ -1,21 +1,46 @@
+from common import Storage
 
-# Function to load data in models/filename to spanner DB
 def load_data(filename, model_name, category):
```
Take the Spanner instance and db as args?

ShobhitBehl replied:
Spanner instance and db name are arguments to Storage class, constant strings in load_data() in main.py files.

rachitagrawal commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,142 @@
+from common import Edge
+from common import Graph
+import OpToNode
+import TensorToEdge
```
nit: import ordering incorrect. 

import should be lexicographically sorted according to each module's full package path.

rachitagrawal commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,142 @@
+
+    # Reading tflite file onto Model object
+    def parse(self, file_path):
+        with open(file_path, "rb") as file:
```
What if invalid file path?

ShobhitBehl replied:
Program terminates with no such file or directory error.

rachitagrawal commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,142 @@
+        
+        # Populating to_nodes, from_nodes
+        # Add proxy nodes for Input and Output of the model
+
```
nit: redundant blank line.

rachitagrawal commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,142 @@
+        from_nodes = dict()
+        
+        # Adding tensors to list of edges
+        offset_edges = len(edges) 
```
Will it not be 0 always?

rachitagrawal commented on TFLite/src/TFLiteParser.py:
```diff
@@ -0,0 +1,142 @@
+            nodes.append(new_node)
+            node_index = len(nodes) - 1
+
+            
```
nit: extra blank line.

rachitagrawal commented on TFLite/src/TensorToEdge.py:
```diff
@@ -0,0 +1,29 @@
+        shape = list()
+        for i in range(tensor.ShapeLength()):
+            shape.append(tensor.Shape(i))
+        new_edge.shape = shape
```
Just a thought, will we be able to query base on dimensions of tensors? Basically, there are use cases when we want to find out how many > 4D tensors are present in the model. If not, then see if that can be added.

ShobhitBehl replied:
Shape stored as ARRAY in spanner, it provides ARRAY functions for querying like ARRAY_LENGTH which will correspond to number of dimensions

rachitagrawal commented on TFLite/src/init.py:
```diff
@@ -0,0 +1,30 @@
+
+    # Absolute path to schema.fbs and the path to store the generated code
+    schema_path = '/home/shobhitbehl/ml-model-characterization/TFLite/schema.fbs'
+    out_path = '/home/shobhitbehl/ml-model-characterization/TFLite/src/'
```
Looks like the change is missed here.

rachitagrawal commented on TFLite/src/main.py:
```diff
@@ -0,0 +1,56 @@
+    duplicated_edges = 0
+
+    for file in os.listdir("models/"):
+        graph = tflite_parser.parse_graph("models/" + file)
```
Let's have models directory as CLA with default value "models".

rachitagrawal commented on TFLite/src/main.py:
```diff
@@ -0,0 +1,56 @@
+        for node in graph.nodes:
+            key = node.serialize()
+
+            if key not in node_vis:
```
nit: node_visited. Same for other variables.

ShobhitBehl replied:
Fixed in later PR.

rachitagrawal reviewed (commented):
Added some comments.

ShobhitBehl commented:
> Added some comments.

Most of the fixes are already present in 'TF Parsing' PR, will reflect upon merge.
I'll push a fix for the rest in the same.
